01-overview-of-ml-systems.md

```diff
@@ -0,0 +1,110 @@
+
+![state-of-enterprise-ml](01-overview-of-ml-systems.assets/state-of-enterprise-ml.png)
+
+
+
+# ML in research vs production
+Research ML and production ML are different. It is important to be aware of the differences to avoid the pitfalls that arise from trying to use a research mentality to production models.
+
+|                        | Research                                                     | Production                                             |
+| ---------------------- | ------------------------------------------------------------ | ------------------------------------------------------ |
+| Requirements           | State-of-the-art model performance with benchmark datasets (1) | Different stakeholders have different requirements (2) |
+| Computational priority | Fast training, high query throughput (3)                     | Fast inference with low latency (4) (5) (6)            |
+| Data                   | Often Static                                                 | Constantly shifting                                    |
+| Fairness               | Often not a focus                                            | Must be considered (7)                                 |
+| Interpretability       | Often not a focus, performance is the north start (8)        | Must be considered                                     |
+
+(1) No focus on model deployment and maintenance.
+
+(2) MLEs, Sales teams, Product Managers, ML platform teams all have different requirements for the same system.
+
+(3) Most modern systems batch up queries at inference time. In research, it doesn't matter if the time to answer a single query is high as long as the throughput of predictions for the whole test dataset is acceptable. In production, the time to answer a single query is important, so we need to balance that with the throughput.
+
+(4)  `...once you deploy your model into the real world, latency matters a lot. In 2017, an Akamai study found that a 100 ms delay can hurt conversion rates by 7%. 20 In 2019, Booking.com found that an increase of about 30% in latency cost about 0.5% in conversion rates—“a relevant cost for our business.” 21 In 2016, Google found that more than half of mobile users will leave a page if it takes more than three seconds to load. 22 Users today are even less patient.` 
+
+(5) Latency must be monitored using percentiles, not averages. Don't ignore higher percentiles like p90 through p99 as the most valuable customers, tend to be the ones with more data and the ones that experience slower inferences.
+
+(6) Consider using high percentiles like p90 to specify the latency requirements of the system.
+
+(7) ML systems encode the past and perpetuate the biases the training data has. To make things worse, ML systems can discriminate at scale potentially affecting millions of people. 
+
+(8) For example, *ensemble models* are common in research but very uncommon in production because of their low interpretability and high complexity.
```

Comment:
Amazon fulfillment always had ensemble models and meta models. It's more complex but totally doable.